mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java

```
@@ -60,25 +69,51 @@
-                
-                attachments.add(Attachment.builder()
-                        .bytes(getBytes(messageWriter, entity.getBody()))
-                        .type(contentType.or(DEFAULT_CONTENT_TYPE))
-                        .name(name)
-                        .build());
+            if (entity.isMultipart()) {
+                attachments.addAll(listAttachments((Multipart) entity.getBody()));
+            } else {
+                if (isAttachment(entity)) {
+                    attachments.add(retrieveAttachment(messageWriter, entity));
+                }
             }
         }
         return attachments.build();
     }
 
+    private MessageAttachment retrieveAttachment(MessageWriter messageWriter, Entity entity) throws IOException {
+        Optional<ContentTypeField> contentTypeField = getContentTypeField(entity);
+        Optional<String> contentType = contentType(contentTypeField);
+        Optional<String> name = name(contentTypeField);
+        Optional<String> cid = cid(castField(entity.getHeader().getField(CONTENT_ID), ContentIdField.class));
+        boolean isInline = isInline(castField(entity.getHeader().getField(CONTENT_DISPOSITION), ContentDispositionField.class));
+
+        return MessageAttachment.builder()
+                .attachment(Attachment.builder()
+                    .bytes(getBytes(messageWriter, entity.getBody()))
+                    .type(contentType.or(DEFAULT_CONTENT_TYPE))
+                    .name(name)
+                    .build())
+                .cid(cid.orNull())
+                .isInline(isInline)
+                .build();
+    }
+
+    private Optional<ContentTypeField> getContentTypeField(Entity entity) {
+        return castField(entity.getHeader().getField(CONTENT_TYPE), ContentTypeField.class);
+    }
+
+    @SuppressWarnings("unchecked")
+    private <U extends ParsedField> Optional<U> castField(Field field, Class<U> clazz) {
+        if (field == null || !clazz.isInstance(field)) {
```

Should we fail when an header is invalid, i'm not sure.